File: chaldean_V2_fileinput.py
import sys
from collections import defaultdict

# Chaldean numerology values
CHALEDEAN_VALUES = defaultdict(lambda: 0, {
    ' ': 0, "a": 1, "b": 2, "c": 3, "d": 4, "e": 5, "f": 8, "g": 3, "h": 5,
    "i": 1, "j": 1, "k": 2, "l": 3, "m": 4, "n": 5, "o": 7, "p": 8, "q": 1,
    "r": 2, "s": 3, "t": 4, "u": 6, "v": 6, "w": 6, "x": 5, "y": 1, "z": 7
})


def reduce_to_single_digit(n: int) -> int:
    """Reduces a number to a single digit (Cheiro method)."""
    while n >= 10:
        n = sum(int(digit) for digit in str(n))
    return n


def reduce_for_linda_goodman(n: int) -> int:
    """Reduces a number for Linda Goodman method, stops at 11 or 22."""
    while n not in (11, 22) and n >= 10:
        n = sum(int(digit) for digit in str(n))
    return n


def calculate_chaldean_parts(name_parts: list[str]) -> list[int]:
    """Returns Chaldean totals for each name part."""
    return [sum(CHALEDEAN_VALUES[char] for char in part) for part in name_parts]


def calculate_linda_goodman_parts(name_parts: list[str]) -> list[int]:
    """Returns Linda Goodman reduced totals per name part."""
    return [
        reduce_for_linda_goodman(sum(CHALEDEAN_VALUES[char] for char in part))
        for part in name_parts
    ]


def calculate_cheiro_parts(name_parts: list[str]) -> list[int]:
    """Returns Cheiro reduced totals per name part."""
    return [
        reduce_to_single_digit(sum(CHALEDEAN_VALUES[char] for char in part))
        for part in name_parts
    ]


def format_name(name: str) -> str:
    """Capitalizes each part of the name nicely."""
    return ' '.join(word.capitalize() for word in name.strip().split())


def process_name(name_input: str):
    """Process and output the numerology for a single name."""
    name_parts = name_input.split()
    formatted_name = format_name(name_input)

    # Part-wise totals
    chaldean_parts = calculate_chaldean_parts(name_parts)
    linda_parts = calculate_linda_goodman_parts(name_parts)
    cheiro_parts = calculate_cheiro_parts(name_parts)

    # Final totals
    chaldean_total = sum(chaldean_parts)
    linda_total = sum(linda_parts)
    cheiro_total = sum(cheiro_parts)

    # Output with aligned labels
    print(f"\nNumerology totals for: {formatted_name}")
    print(f"{'Chaldean Numerology'.ljust(24)}: {chaldean_parts} & {chaldean_total}")
    print(f"{'Linda Goodman Numerology'.ljust(24)}: {linda_parts} & {linda_total}")
    print(f"{'Cheiro Numerology'.ljust(24)}: {cheiro_parts} & {cheiro_total}")


def main():
    # If the -f flag is provided
    if len(sys.argv) > 2 and sys.argv[1] == '-f':
        file_name = sys.argv[2]
        try:
            with open(file_name, 'r') as file:
                names = file.readlines()
            # Process each name in the file
            for name in names:
                name = name.strip()  # remove extra whitespace
                if name:  # if the name is not empty
                    process_name(name)
        except FileNotFoundError:
            print(f"Error: The file '{file_name}' was not found.")
    else:
        # If no -f flag, ask the user for a single name
        name_input = input("Please enter your name: ").strip().lower()
        if not name_input:
            print("Name cannot be empty. Please try again.")
        else:
            process_name(name_input)


if __name__ == "__main__":
    main()

File: chaldean_V2_fileinput.clj
(ns numerology.core
  (:require [clojure.string :as str])
  (:gen-class))

;; Chaldean numerology values
(def chaldean-values
  {"a" 1 "b" 2 "c" 3 "d" 4 "e" 5 "f" 8 "g" 3 "h" 5
   "i" 1 "j" 1 "k" 2 "l" 3 "m" 4 "n" 5 "o" 7 "p" 8
   "q" 1 "r" 2 "s" 3 "t" 4 "u" 6 "v" 6 "w" 6 "x" 5
   "y" 1 "z" 7 " " 0})

(defn char-value [ch]
  (get chaldean-values (str/lower-case (str ch)) 0))

(defn reduce-to-single-digit [n]
  (loop [num n]
    (if (< num 10)
      num
      (recur (reduce + (map #(Character/digit % 10) (str num)))))))

(defn reduce-for-linda-goodman [n]
  (loop [num n]
    (if (or (= num 11) (= num 22) (< num 10))
      num
      (recur (reduce + (map #(Character/digit % 10) (str num)))))))

(defn sum-name-part [part]
  (reduce + (map char-value part)))

(defn chaldean-parts [name-parts]
  (map sum-name-part name-parts))

(defn linda-goodman-parts [name-parts]
  (map (comp reduce-for-linda-goodman sum-name-part) name-parts))

(defn cheiro-parts [name-parts]
  (map (comp reduce-to-single-digit sum-name-part) name-parts))

(defn format-name [name]
  (->> (str/split name #"\s+")
       (map #(str/capitalize (str/trim %)))
       (str/join " ")))

(defn process-name [name-input]
  (let [name-parts (str/split (str/lower-case name-input) #"\s+")
        formatted (format-name name-input)
        chaldean (chaldean-parts name-parts)
        linda (linda-goodman-parts name-parts)
        cheiro (cheiro-parts name-parts)
        total-chaldean (reduce + chaldean)
        total-linda (reduce + linda)
        total-cheiro (reduce + cheiro)]
    (println (str "\nNumerology totals for: " formatted))
    (println (format "%-24s: %s & %d" "Chaldean Numerology" chaldean total-chaldean))
    (println (format "%-24s: %s & %d" "Linda Goodman Numerology" linda total-linda))
    (println (format "%-24s: %s & %d" "Cheiro Numerology" cheiro total-cheiro))))

(defn -main [& args]
  (cond
    (and (= (first args) "-f") (second args))
    (let [file-name (second args)]
      (try
        (let [lines (->> (slurp file-name) str/split-lines (remove str/blank?))]
          (doseq [line lines] (process-name line)))
        (catch java.io.FileNotFoundException _
          (println (str "Error: The file '" file-name "' was not found.")))))

    :else
    (do
      (print "Please enter your name: ") (flush)
      (let [input (str/lower-case (read-line))]
        (if (str/blank? input)
          (println "Name cannot be empty. Please try again.")
          (process-name input))))))
